apps/pp/utils.py

```diff
@@ -23,19 +35,85 @@ def get_relationship_id(root_serializer, name):
     return val
 
 
-def get_resource_name(model, attr=None):
-    if attr not in (None, 'pk', 'id'):
-        model = model._meta.get_field(attr).related_model
-    return model.JSONAPIMeta.resource_name
+def get_jsonapimeta(obj, related_field=None):
+    model = obj
+    if isinstance(obj, models.Model):
+        if related_field and related_field not in ('pk', 'id'):
+            model = obj._meta.get_field(related_field).related_model
+    elif isinstance(obj, models.QuerySet):
+        model = obj.model
+    return model.JSONAPIMeta
+
+
+def get_resource_name(obj, related_field=None, model=None):
+    meta = get_jsonapimeta(model if model else obj, related_field)
+    name = getattr(meta, 'resource_name', None)
+    if name is not None:
+        return name
+    get_names = getattr(meta, 'get_resource_names', None)
+    if get_names is not None:
+        return get_names(obj)
+    raise ValueError('obj need to have either resource_name or get_resource_names defined')
+
```

Comment:
I think a sentence of comment (argument type?) or a more descriptive name for both `get_jsonapimeta` and `get_resource_name` would be nice (unless the description would be as complicated as the function itself :) )
"why would I get a json_api_meta? resource name of what?"

Reply:
This is just helper of helper to make structure of calls neat, it does dirty job to get jsonapi more universally. I guess you do not need to use that, unless you refactor utils, it's kind of private, but not strictly, so no "_" before name.